AK/FloatingPointStringConversions.cpp

```diff
@@ -0,0 +1,2018 @@
+
+static i32 decimal_exponent_to_binary_exponent(i32 exponent)
+{
+    return ((((152170 + 65536) * exponent) >> 16) + 63);
+}
+
+static u128 multiply(u64 a, u64 b)
+{
+#ifdef __SIZEOF_INT128__
+    unsigned __int128 result = (unsigned __int128)a * b;
+    u64 low = result;
+    u64 high = result >> 64;
+    return u128 { low, high };
+#else
+    return u128 { a }.wide_multiply(u128 { b }).low;
+#endif
+}
+
+template<unsigned Precision>
+u128 multiplication_approximation(u64 value, i32 exponent)
+{
+    auto z = power_of_five(exponent);
+
+    static_assert(Precision < 64);
+    constexpr u64 mask = NumericLimits<u64>::max() >> Precision;
+
+    auto lower_result = multiply(z.high(), value);
+
+    if ((lower_result.high() & mask) == mask) {
+        auto upper_result = multiply(z.low(), value);
+        lower_result.low() += upper_result.high();
+        if (upper_result.high() > lower_result.low()) {
+            ++lower_result.high();
+        }
+    }
+
+    return lower_result;
+}
+
+template<typename T>
+static FloatingPointBuilder not_enough_precision_binary_to_decimal(i64 exponent, u64 mantissa, int leading_zeros)
+{
+    using FloatingPointRepr = FloatingPointInfo<T>;
+    i32 did_not_have_upper_bit = static_cast<i32>(mantissa >> 63) ^ 1;
+    FloatingPointBuilder answer;
+    answer.mantissa = mantissa << did_not_have_upper_bit;
+
+    i32 bias = FloatingPointRepr::mantissa_bits() + FloatingPointRepr::exponent_bias();
+    answer.exponent = decimal_exponent_to_binary_exponent(static_cast<i32>(exponent)) - leading_zeros - did_not_have_upper_bit - 62 + bias;
+    // Make it negative to show we need more precision.
+    answer.exponent -= FloatingPointBuilder::invalid_exponent_offset;
+    VERIFY(answer.exponent < 0);
+    return answer;
+}
+
+template<typename T>
+static FloatingPointBuilder fallback_binary_to_decimal(u64 mantissa, i64 exponent)
+{
+    // We should have caught huge exponents already
+    VERIFY(exponent >= -400 && exponent <= 400);
+
+    // Perform the initial steps of binary_to_decimal.
+    auto w = mantissa;
+    auto leading_zeros = count_leading_zeroes(mantissa);
+    w <<= leading_zeros;
+
+    auto product = multiplication_approximation<FloatingPointInfo<T>::mantissa_bits() + 3>(w, exponent);
+
+    return not_enough_precision_binary_to_decimal<T>(exponent, product.high(), leading_zeros);
+}
+
+template<typename T>
+static FloatingPointBuilder binary_to_decimal(u64 mantissa, i64 exponent)
+{
+    using FloatingPointRepr = FloatingPointInfo<T>;
+
+    if (mantissa == 0 || exponent < FloatingPointRepr::min_power_of_10())
+        return FloatingPointBuilder::zero();
+
+    // Max double value which isn't negative is xe308
+    if (exponent > FloatingPointRepr::max_power_of_10())
+        return FloatingPointBuilder::infinity<T>();
+
+    auto w = mantissa;
+    // Normalize the decimal significand w by shifting it so that w ∈ [2^63 , 2^64)
```

Comment:
```suggestion
    // Normalize the decimal significand w by shifting it so that w ∈ [2^63, 2^64]
```

Reply:
This is not a mismatched bracket, it means the range from 2^63 included to 2^64 excluded

The space before the comma I will fix though

Reply:
I guess I can remove the comment as well if you think it's confusing

Reply:
TIL! I've always been more of a programming person than a math person, so I intuitively see it as a typo, but it definitely seems to be common range notation.